### ISSUE TITLE
Track piece ID 95 becomes 257 upon load if track type was altered

### ISSUE THREAD

**Issue by sta5is** [OpenRCT2/OpenRCT2#15258]
### Operating System

Windows 10, 64-bit

### OpenRCT2 build

OpenRCT2, v0.3.4.1 (9511104 on develop) provided by GitHub

### Describe the issue

Track piece ID 95 (upward steep left turn) becomes ID 257 (beyond limits) upon loading into the game

![track piece 95](https://user-images.githubusercontent.com/26368692/130336804-c745dedc-3577-469f-af6a-0498845ed12e.PNG)

![track piece 257](https://user-images.githubusercontent.com/26368692/130336805-48782e80-c7b7-48f0-ba92-f5b16d4f287a.PNG)

### Steps to reproduce

1. Build upward steep left turn track piece (ID 95)
2. Change track type to a flat ride (for example: carousel)
3. Save session
4. Upon load, track ID 95 becomes 257

### Attachments

Affected save can be provided via private message because the park is being built for an ongoing competition

**Comment by spacek531** [comment]
Does this happen if you set the ride type to Lift?

**Comment by sta5is** [comment]
No, setting to lift/elevator type and then save/reload retains the track ID as 95 correctly.
Save example:
[track id 95 problem.zip](https://github.com/OpenRCT2/OpenRCT2/files/7027745/track.id.95.problem.zip)

**Comment by duncanspumpkin** [comment]
This is expected behaviour.

**Comment by spacek531** [comment]
These are all the track pieces that get overwritten with different track types when you change the ride type to a flat ride:

LeftQuarterTurn1TileUp60
Up25LeftBanked
Up25RightBanked
Down25LeftBanked
Down25RightBanked
FlatToUp60LongBase
Up60ToFlatLongBase
Down60ToFlatLongBase
FlatToDown60LongBase
CableLiftHill

Lift is a tracked ride (as is vertical freefall tower, roto-drop tower, observation tower), so they will not have this behavior. You have to set Lift before the track piece gets overwritten, not after.

**Comment by sta5is** [comment]
Thank you for the clarification. I usually use flat ride ride types because they generally generate more favorable ratings. May I ask why this happens with the left steep 1 tile upwards turn but not the right piece counterpart?

**Comment by duncanspumpkin** [comment]
CS in his infinite wisdom decided to use the same ids for flat and tracked rides. This meant that every single time you tried to use a track you had to check to see if it was a flat ride or not. This made for a lot of repetitive code. So instead we assign it a different track id so you don't need to know if it's flat or not. But due to sv6 export we need to undo that on import/export.

**Comment by ocalhoun6** [comment]
So this could potentially be solved in the new save format?

**Comment by duncanspumpkin** [comment]
Yes in the nsf it doesn't do import/export.

**Comment by spacek531** [comment]
> Furthermore, adding to your list @spacek531 , quarter loops get overwritten too when the track type is changed to a flat ride type

Can you tell me the IDs of the pieces that get overwritten?

**Comment by Gymnasiast** [comment]
This issue is related: https://github.com/OpenRCT2/OpenRCT2/issues/14482

In general, I do think the heuristics could be improved. It should perhaps take a look at the vehicle, and only convert track piece IDs if the vehicle belongs to a flat ride as well.

Additionally, we should really have an invisibility bit for rides in the NSF so that flat ride hacks are no longer needed (for new parks, of course).

**Comment by duncanspumpkin** [comment]
https://gist.github.com/duncanspumpkin/90f361690e029b8ad2acea3ccb613dea this was my attempt at an alternative conversion. But I had forgotten that we also have vehicles on the ride doing conversions. I think its fine that they do not perform any conversion (i.e. delete the current conversion code on them). @Gymnasiast  any thoughts?
